open_instruct/ppo_fast.py

```diff
@@ -0,0 +1,1866 @@
+
+        ds_config = get_train_ds_config(
+            offload=False,
+            adam_offload=False,
+            stage=args.deepspeed_stage,
+            bf16=True,
+        )
+        ds_config["train_micro_batch_size_per_gpu"] = args.per_device_train_batch_size
+        ds_config["gradient_accumulation_steps"] = 1
+        # @vwxyzjn: MAGIC: it's actually needed to initialize this `dschf`, so
+        # https://huggingface.co/docs/transformers/deepspeed#non-trainer-deepspeed-integration
+        # next line instructs transformers to partition the model directly over multiple gpus using
+        # deepspeed.zero.Init when model's `from_pretrained` method is called.
+        if ds_config is not None and ds_config["zero_optimization"]["stage"] == 3:
+            HfDeepSpeedConfig(ds_config)
+        else:
+            pass
+
+        self.policy: PreTrainedModel = AutoModelForCausalLM.from_pretrained(
+            model_config.model_name_or_path,
+            revision=model_config.model_revision,
+            torch_dtype=torch.bfloat16,
+            attn_implementation="flash_attention_2",
+            use_cache=False,
+        )
+        disable_dropout_in_model(self.policy)
+        self.policy.gradient_checkpointing_enable()
+        from deepspeed.ops.adam import FusedAdam  # DeepSpeedCPUAdam
+
+        if args.set_weight_decay_on_bias_and_norm:
+            optim_params = get_optimizer_grouped_parameters(self.policy, args.weight_decay)
+        else:
+            optim_params = self.policy.parameters()
+        self.optimizer = FusedAdam(
+            optim_params, lr=args.learning_rate, betas=(0.9, 0.95), weight_decay=args.weight_decay
+        )
+        num_scheduler_steps = args.num_training_steps * args.num_epochs * args.num_mini_batches
+        warm_up_steps = args.warm_up_steps
+        if args.warmup_ratio > 0.0:
+            warm_up_steps = int(num_scheduler_steps * args.warmup_ratio)
+        scheduler = get_scheduler(
+            args.lr_scheduler_type,
+            optimizer=self.optimizer,
+            num_warmup_steps=warm_up_steps,
+            num_training_steps=num_scheduler_steps,
+        )
+        self.model, self.optimizer, _, self.scheduler = deepspeed.initialize(
+            model=self.policy,
+            optimizer=self.optimizer,
+            config=ds_config,
+            lr_scheduler=scheduler,
+            dist_init_required=True,
+        )
+        self.model.train()
+
+        # value model
+        from open_instruct.olmo_adapter import (
+            Olmo2Config,
+            Olmo2ForSequenceClassification,
+            OlmoeConfig,
+            OlmoeForSequenceClassification,
+        )
+
+        AutoModelForSequenceClassification.register(Olmo2Config, Olmo2ForSequenceClassification)
+        AutoModelForSequenceClassification.register(OlmoeConfig, OlmoeForSequenceClassification)
+        self.value_model: PreTrainedModel = AutoModelForSequenceClassification.from_pretrained(
```

Comment:
Out of scope for this PR, but I wonder if we can get more creative with the value model, like re-using the LM head, or loading pretrained value models...